### PR TITLE
cli: Add `check:theia-extensions` command

### PR DESCRIPTION
**Pull request by sgraband**
#### What it does

The goal of this PR is to improve the existing `theia check:dependencies` script introduced in [this PR](https://github.com/eclipse-theia/theia/pull/11483). The existing script returns a list of all mismatching version numbers for all dependencies. While introducing two versions of the same dependency can always possibly introduce issues, this will be much more common for Theia extensions, than for other dependencies like `loadash` or `rimraf`. Thats why:

This allows to check for duplicate versions of theia extensions only. 
Added the option `only-theia-extensions` to `theia check:dependencies`. 
Theia extensions are identified via their `theiaExtensions` field in the `package.json`. 
The check is running recursively as long as it finds Theia extensions.
The recursion was added to also display cases where more than 2 versions of a dependency might extist.
The advantage of this approach is that a shorter, but more important, list is returned.

Fixes # 12572.

Contributed on behalf of STMicroelectronics
#### How to test

- Check out this PR and build with `yarn`
- Go to `dev-packages/cli` and run `yarn link`
- Checkout [this example](https://github.com/eclipsesource/theia/tree/cliExampleApplication) (or create any other example you would like) and run `yarn link "@theia/cli"` in its root
- Run `yarn` to see the output (this will run `theia check:theia-extensions` on postinstall)

The provided example has a few issues built-in:
- The `test-extension` requires a `@theia/core` version of `1.37.2` while the application depends on `1.38.0`
- To test nested (and 3rd-party)  dependencies the following dependencies where added
    - `test-extension` > `test-extension-3` > `@eclipse-emfcloud/modelserver-theia:0.8.0-theia-cr03`
    - `test-extension2` > `test-extension-4` > `@eclipse-emfcloud/modelserver-theia:0.8.0-theia-cr01`
    - `@eclipse-emfcloud/modelserver-theia:0.8.0-theia-cr01` depends on version `1.37.2` of `@theia/(core|filesystem|process|variable-resolver|workpace)`

So the result of `theia check:theia-extensions` should find the following issues:
- Different versions for `@theia/(filesystem|process|variable-resolver|workpace)` from the application and `@eclipse-emfcloud/modelserver-theia`
- Different versions for `@theia/core` from the application, `@eclipse-emfcloud/modelserver-theia` and `test-extension`
- Different versions for `@eclipse-emfcloud/modelserver-theia` from `test-extension-3` and `test-extension-4`

Note that one of the versions will always be hoisted to the root `node_modules` version

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: dev-packages/cli/src/check-dependencies.ts [hunk 1]
```diff
@@ -122,9 +125,18 @@ function findDependencies(workspace: string, options: CheckDependenciesOptions):
...
-            matchingPackageJsons.push(toDependency(packageJson, nodeModulesDir, dependent))
-        )
+        }).forEach(packageJsonPath => {
+            const dependency = toDependency(packageJsonPath, nodeModulesDir, dependent);
+            if (!options.onlyTheiaExtensions) {
+                matchingPackageJsons.push(dependency);
+            } else {
+                if (dependency.isTheiaExtension) {
+                    matchingPackageJsons.push(dependency);
+                    const childNodeModules: string = path.join(nodeModulesDir, packageJsonPath, `../${NODE_MODULES}`);
```
Comment (tsmaeder) [inline dev-packages/cli/src/check-dependencies.ts:135]: Why do we need to find the transitive dependencies "by hand" here?

File: dev-packages/cli/src/check-dependencies.ts [hunk 2]
```diff
@@ -133,12 +145,14 @@ function toDependency(packageJsonPath: string, nodeModulesDir: string, dependent
     const fullPackageJsonPath = path.join(process.cwd(), nodeModulesDir, packageJsonPath);
     const name = getPackageName(fullPackageJsonPath);
     const version = getPackageVersion(fullPackageJsonPath);
+    const theiaExtension = isTheiaExtension(fullPackageJsonPath);
```
Comment (tsmaeder) [inline dev-packages/cli/src/check-dependencies.ts:148]: Why not just inline this one? Otherwise, "isTheiaExtension" would be more descriptive.

File: dev-packages/cli/src/theia.ts [hunk 3]
```diff
@@ -280,6 +309,12 @@ async function theiaCli(): Promise<void> {
                     boolean: true,
                     default: false
                 },
+                'only-theia-extensions': {
+                    alias: 'o',
+                    describe: 'Only check theia extensions if they are resolved to a single version',
```
Comment (tsmaeder) [inline dev-packages/cli/src/theia.ts:314]: This is confusing: do you mean "only execute the check if all theia extensions resolve to a single version" or "check whether all theia extensions resolve to the same, unique version" or "check whether each time, a theia extension version is resolved, it's the same version for the same package". 🤷

Comment (tsmaeder) [reply dev-packages/cli/src/theia.ts:314]: Or "only check packages which are theia extensions"?

File: dev-packages/cli/src/check-dependencies.ts (hunk 1 above)
Comment (sgraband) [reply dev-packages/cli/src/check-dependencies.ts]: So as far as i can tell, the old version simply looked at the first level of `node_modules` in root and all workspaces specified. To quickly test, i ran the check for all dependencies (not only theia extensions) and it actually does not show the versions mismatch found in `node_modules/@eclipse-emfcloud/modelserver-theia/node_modules/@theia/core/package.json`. So maybe this was simply overlooked?
I can adjust the code to recursively call `findDependencies` for each of the three use cases.

Comment (tsmaeder) [reply dev-packages/cli/src/check-dependencies.ts]: Either way, methinks it makes sense to either always recurse or never.

Comment (sgraband) [reply dev-packages/cli/src/check-dependencies.ts]: I changed it so that all checks run recursively. To make sure that only package.json files in the root of a package are picked up i kept the "by hand" way instead of changing the glob statement. I guess the code is easier to read like this. Or would you say, that we should adjust the glob to do all of this and save the recursion?

Comment (sgraband) [reply dev-packages/cli/src/check-dependencies.ts]: @tsmaeder any objections on opening the upstream PR?

Comment (planger) [reply dev-packages/cli/src/check-dependencies.ts]: I haven't looked into the requirements this PR aims to address yet. For now I can just say that -- as far as I remember -- we didn't need a recursive check before, because `yarn install` already does the recursive resolution for us and puts the result into the `node_modules` folders. Thus checking the `node_modules` folders is enough as it is the flattened list of all direct and indirect dependencies.

Comment (planger) [reply dev-packages/cli/src/check-dependencies.ts]: Yes you are right!

The current version on master only checks the root level and first level (where extensions and browser-app/electron-app typically are). If the dependencies are hoisted, however, checking the root level and the first level should imho always be already sufficient to at least identify that we've pulled multiple versions of the same dependency, because hoisting always starts from root to leaf, so at least one version would already be on root and one would be on level 1.

So turning this into a recursive check will imho not identify additional error cases, but enables giving more complete error messages; that is, it enables listing all pulled versions of a dependency, not just the first two versions (that happened to be in root and level 1). For non-hoisted cases, the situation may be completely different though -- I haven't tested with such scenarios yet.

In summary, if we want to be able to provide a complete error message (without having to run `yarn why` after the error was printed), the recursive check makes sense I guess.

Also keeping both checks `theia check:theia-version` (a) and your new one (b) makes sense, because (b) checks for uniqueness of packages, whereas (a) checks whether the exact same Theia version is used across all dependencies (also if they are unique but different).
